drivers/tee/tee_shm.c

```diff
@@ -20,25 +20,42 @@
 #include <linux/tee_drv.h>
 #include "tee_private.h"
 
+/* extra references appended to shm object for registered shared memory */
+struct dmabuf_ref {
+	struct dma_buf *dmabuf;
+	struct dma_buf_attachment *attach;
+	struct sg_table *sgt;
+};
+
 static void tee_shm_release(struct tee_shm *shm)
 {
 	struct tee_device *teedev = shm->teedev;
-	struct tee_shm_pool_mgr *poolm;
 
 	mutex_lock(&teedev->mutex);
 	idr_remove(&teedev->idr, shm->id);
 	if (shm->ctx)
 		list_del(&shm->link);
 	mutex_unlock(&teedev->mutex);
 
-	if (shm->flags & TEE_SHM_DMA_BUF)
-		poolm = &teedev->pool->dma_buf_mgr;
-	else
-		poolm = &teedev->pool->private_mgr;
+	if (shm->flags & TEE_SHM_MAPPED) {
+		struct tee_shm_pool_mgr *poolm;
 
-	poolm->ops->free(poolm, shm);
-	kfree(shm);
+		if (shm->flags & TEE_SHM_DMA_BUF)
+			poolm = &teedev->pool->dma_buf_mgr;
+		else
+			poolm = &teedev->pool->private_mgr;
+
+		poolm->ops->free(poolm, shm);
+	} else {
+		struct dmabuf_ref *ref = (struct dmabuf_ref *)(shm + 1);
```

Comment:
So without `TEE_SHM_MAPPED` this is always what we would get? It would feel a bit safer to have something more than the absence of a flag to indicate that there's a `struct dmabuf_ref` following.